scipy/stats/_distribution_infrastructure.py

```diff
@@ -0,0 +1,5082 @@
+
+        If a formula for the mode is not specifically implemented for the
+        chosen distribution, SciPy will attempt to compute the mode
+        numerically, which may not meet the user's preferred definition of a
+        mode. In such cases, the user is encouraged to subclass the
+        distribution and override ``mode``.
+
+        Examples
+        --------
+        Instantiate a distribution with the desired parameters:
+
+        >>> from scipy import stats
+        >>> X = stats.Normal(mu=1., sigma=2.)
+
+        Evaluate the mode:
+
+        >>> X.mode()
+        1.0
+
+        If the mode is not uniquely defined, ``mode`` nonetheless returns a
+        single value.
+
+        >>> X = stats.Uniform(a=0., b=1.)
+        >>> X.mode()
+        0.5
+
+        If this choice does not satisfy your requirements, subclass the
+        distribution and override ``mode``:
+
+        >>> class BetterUniform(stats.Uniform):
+        ...     def mode(self):
+        ...         return self.b
+        >>> X = BetterUniform(a=0., b=1.)
+        >>> X.mode()
+        1.0
+
+        """
+        return self._mode_dispatch(method=method, **self._parameters)
+
+    @_dispatch
+    def _mode_dispatch(self, method=None, **kwargs):
+        # We could add a method that looks for a critical point with
+        # differentiation and the root finder
+        if self._overrides('_mode_formula'):
+            method = self._mode_formula
+        else:
+            method = self._mode_optimization
+        return method
+
+    def _mode_formula(self, **kwargs):
+        raise NotImplementedError(self._not_implemented)
+
+    def _mode_optimization(self, **kwargs):
+        # Heuristic until we write a proper minimization bracket finder (like
+        # bracket_root): if the PDF at the 0.01 and 99.99 percentiles is not
+        # less than the PDF at the median, it's either a (rare in SciPy)
+        # bimodal distribution (in which case the generic implementation will
+        # never be great) or the mode is at one of the endpoints.
+        if not np.prod(self._shape):
+            return np.empty(self._shape, dtype=self._dtype)
+        p_shape = (3,) + (1,)*len(self._shape)
+        p = np.asarray([0.0001, 0.5, 0.9999]).reshape(p_shape)
+        bracket = self._icdf_dispatch(p, **kwargs)
+        f, args = _kwargs2args(lambda x, **kwargs: -self._pdf_dispatch(x, **kwargs),
+                               args=(), kwargs=kwargs)
+        res = _chandrupatla_minimize(f, *bracket, args=args)
+        mode = np.asarray(res.x)
+        mode_at_boundary = ~res.success
+        mode_at_left = mode_at_boundary & (res.fl <= res.fr)
+        mode_at_right = mode_at_boundary & (res.fr < res.fl)
+        a, b = self._support(**kwargs)
+        mode[mode_at_left] = a[mode_at_left]
+        mode[mode_at_right] = b[mode_at_right]
+        return mode[()]
+
+    def mean(self, *, method=None):
+        r"""Mean (raw first moment about the origin)
```

Comment:
I would put the details about moment in a note section (same for the others) and point to `moment` for more details. I suspect that would be distracting otherwise.

Reply:
I only added it because it seemed strange to have a one word summary that was identical to the name of the function.

If I were to move it from this line, it would become the extended summary.

```python3
    def mean(self, *, method=None):
        r"""Mean

        Raw first moment about the origin.
```
Let's see what the others think.

Reply:
I like it better as is. Having the one word summary seems off to me.